MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BufferProcessor.java

```diff
@@ -0,0 +1,128 @@
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import static android.os.Build.VERSION_CODES.LOLLIPOP;
+
+/**
+ * A handler meant specifically for handling messages received from a motion photo reader.
+ */
+class BufferProcessor {
+    private static final long TIMEOUT_US = 1000L;
+    private static final long US_TO_NS = 1000L;
+
+    /**
+     * Fields shared with motion photo reader.
+     */
+    private final MediaExtractor lowResExtractor;
+    private final MediaCodec lowResDecoder;
+    private final BlockingQueue<Integer> availableInputBuffers;
+    private final BlockingQueue<Bundle> availableOutputBuffers;
+
+    public BufferProcessor(MediaExtractor lowResExtractor, MediaCodec lowResDecoder,
+                           BlockingQueue<Integer> availableInputBuffers, BlockingQueue<Bundle> availableOutputBuffers) {
+        this.lowResExtractor = lowResExtractor;
+        this.lowResDecoder = lowResDecoder;
+        this.availableInputBuffers = availableInputBuffers;
+        this.availableOutputBuffers = availableOutputBuffers;
+    }
+
+    /**
+     * Retrieve the index of the next available input buffer.
+     * @return the index of the next available input buffer, or -1 if the poll call results in a timeout.
+     */
+    @RequiresApi(api = LOLLIPOP)
+    private int getAvailableInputBufferIndex() {
+        int bufferIndex = -1;
+        try {
+            bufferIndex = availableInputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        return bufferIndex;
+    }
+
+    /**
+     * Read the next data sample from the motion photo to a given input buffer and advance the extractor.
+     */
+    private void readFromExtractor(ByteBuffer inputBuffer, int bufferIndex) {
+        int sampleSize = lowResExtractor.readSampleData(inputBuffer, 0);
+        if (sampleSize < 0) {
+            lowResDecoder.queueInputBuffer(bufferIndex, 0, 0, 0, MediaCodec.BUFFER_FLAG_END_OF_STREAM);
+        }
+        else {
+            lowResDecoder.queueInputBuffer(bufferIndex, 0, sampleSize, lowResExtractor.getSampleTime(), 0);
+            lowResExtractor.advance();
+        }
+    }
+
+    /**
+     * Retrieve the information of the next available output buffer.
+     * @return the bundle containing the information of the next available output buffer, or null if
+     * the poll call results in a timeout.
+     */
+    private Bundle getAvailableOutputBufferData() {
+        Bundle bufferData = null;
+        try {
+            bufferData = availableOutputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        return bufferData;
+    }
+
+    /**
+     * Handle calls to nextFrame() and seekTo() by the MotionPhotoReader.
+     */
+    @RequiresApi(api = LOLLIPOP)
+    public void handleBundle(Bundle messageData) {
```

Comment:
nit: rename to  `process(Bundle messageData)`

also, can we add in comments somewhere (or javadoc of method) what the expected bundle comments need to be ?
e.g.

"MESSAGE_KEY"   - integer containing command 
"TIMESTAMP_US" - ??
"BUFFER_INDEX" - ...